### PR TITLE
chore(tests): build and deploy Lambda Layer stack once 

### LINKED ISSUE #1464: Refactor E2E: Build Lambda Layer once and expose to children stacks

### PR DESCRIPTION
**Pull request by heitorlessa**
**Issue number:** #1464

## Summary

This PR introduces `LambdaLayerStack` that is deployed per test session. It coordinates all parallel workers across CPU to wait for completion before they begin deploying their respective stacks and test run.

This diagram may better explain. More details on a few differences at the comment section.

```mermaid
graph TD
    A[make e2e test] -->Spawn{"Split and group tests <br>by feature and CPU"}

    Spawn -->|Worker0| Worker0_Start["Load tests"]
    Spawn -->|Worker1| Worker1_Start["Load tests"]
    Spawn -->|WorkerN| WorkerN_Start["Load tests"]

    Worker0_Start -->|Wait| LambdaLayerStack["Lambda Layer Stack Deployment"]
    Worker1_Start -->|Wait| LambdaLayerStack["Lambda Layer Stack Deployment"]
    WorkerN_Start -->|Wait| LambdaLayerStack["Lambda Layer Stack Deployment"]

    LambdaLayerStack -->|Worker0| Worker0_Deploy["Launch feature stack"]
    LambdaLayerStack -->|Worker1| Worker1_Deploy["Launch feature stack"]
    LambdaLayerStack -->|WorkerN| WorkerN_Deploy["Launch feature stack"]

    Worker0_Deploy -->|Worker0| Worker0_Tests["Run tests"]
    Worker1_Deploy -->|Worker1| Worker1_Tests["Run tests"]
    WorkerN_Deploy -->|WorkerN| WorkerN_Tests["Run tests"]

    Worker0_Tests --> ResultCollection
    Worker1_Tests --> ResultCollection
    WorkerN_Tests --> ResultCollection

    ResultCollection{"Wait for workers<br/>Collect test results"}
    ResultCollection --> TestEnd["Report results"]
    ResultCollection --> DeployEnd["Delete Stacks"]
```

### Changes

> Please provide a summary of what's being changed

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by heitorlessa** [comment]
I've increased test distribution [parallelization algorithm](https://pytest-xdist.readthedocs.io/en/latest/distribution.html) from module (loadscope) to per test file (loadist). 

* Sequential test: 279.52s
* Parallel test + Layer per stack: 254.53s
* Parallel + Single Layer: 243.58s

Numbers are slightly higher now as I've included all extra dependencies to futureproof when we test Parser.

**Comment by heitorlessa** [comment]
Coordinating across shared queue memory required some changes in how we define infrastructure and their dependencies to run.

`infrastructure` fixtures per feature (e.g., Tracer, Logger, etc.) must accept `layer_arn` argument. This dependency injection also ensures all workers scattered across CPUs will wait for a session fixture. We then moved away from `deploy_once` as this is controlled up a level, making it more explicit (slightly more verbose too :/)

**BEFORE**

```python
from pathlib import Path

import pytest

from tests.e2e.logger.infrastructure import LoggerStack


@pytest.fixture(autouse=True, scope="module")
def infrastructure(request: pytest.FixtureRequest, tmp_path_factory: pytest.TempPathFactory, worker_id: str):
    yield from deploy_once(stack=LoggerStack, request=request, tmp_path_factory=tmp_path_factory, worker_id=worker_id)
```

**AFTER**

```python
from pathlib import Path

import pytest

from tests.e2e.logger.infrastructure import LoggerStack


@pytest.fixture(autouse=True, scope="module")
def infrastructure(request: pytest.FixtureRequest, lambda_layer_arn: str):
    stack = LoggerStack(handlers_dir=Path(f"{request.path.parent}/handlers"), layer_arn=lambda_layer_arn)
    try:
        yield stack.deploy()
    finally:
        stack.delete()
```

**Comment by heitorlessa** [comment]
Stacks changed signature to accept a new optional argument (`layer_arn`). To reduce some level of verbosity I also added a class attribute (`FEATURE_NAME`) to make it easier for newcomers to copy/paste as only `create_resources` and the feature name will change.

**BEFORE**

```python
from pathlib import Path

from tests.e2e.utils.infrastructure import BaseInfrastructureV2


class LoggerStack(BaseInfrastructureV2):
    def __init__(self, handlers_dir: Path, feature_name: str = "logger") -> None:
        super().__init__(feature_name, handlers_dir)

    def create_resources(self):
        self.create_lambda_functions()
```


**AFTER**

```python
from pathlib import Path

from tests.e2e.utils.infrastructure import BaseInfrastructureV2


class LoggerStack(BaseInfrastructureV2):
    FEATURE_NAME = "logger"

    def __init__(self, handlers_dir: Path, feature_name: str = FEATURE_NAME, layer_arn: str = "") -> None:
        super().__init__(feature_name, handlers_dir, layer_arn)

    def create_resources(self):
        self.create_lambda_functions()
```

File: tests/e2e/utils/infrastructure.py [hunk 1]
```diff
@@ -21,11 +19,7 @@
 
 PYTHON_RUNTIME_VERSION = f"V{''.join(map(str, sys.version_info[:2]))}"
 
-
-class PythonVersion(Enum):
-    V37 = {"runtime": Runtime.PYTHON_3_7, "image": Runtime.PYTHON_3_7.bundling_image.image}
-    V38 = {"runtime": Runtime.PYTHON_3_8, "image": Runtime.PYTHON_3_8.bundling_image.image}
-    V39 = {"runtime": Runtime.PYTHON_3_9, "image": Runtime.PYTHON_3_9.bundling_image.image}
```
Comment (rubenfonseca) [inline tests/e2e/utils/infrastructure.py:28]: ❤️

File: tests/e2e/utils/infrastructure.py [hunk 2]
```diff
@@ -38,191 +32,22 @@ def __call__(self) -> Tuple[dict, str]:
...
-                        "${AWS::Region}", region
-                    ),
-                    "asset_path": asset_path,
-                    "asset_packaging": asset_packaging,
-                }
-
-        return assets
-
-    def _transform_output(self, outputs: dict):
-        return {output["OutputKey"]: output["OutputValue"] for output in outputs if output["OutputKey"]}
```
Comment (rubenfonseca) [inline tests/e2e/utils/infrastructure.py:218]: ❤️ ❤️

**Review by rubenfonseca — APPROVED** [review]
Amazing work Heitor, no comments from me. I was expecting a lot of logic to handle the "map-reduce", but it seems the pytest plugin already handles this for you. So no comments here :)

File: tests/e2e/utils/infrastructure.py (hunk 1 above)
Comment (heitorlessa) [reply tests/e2e/utils/infrastructure.py:28]: hmmmm I think this is a git issue, as it's still there?

**Comment by heitorlessa** [comment]
I'm gonna remove the old BaseInfrastructure and rename V2 so we solidify this thing now that all stacks are migrated.

**Comment by heitorlessa** [comment]
thanks for the approval @rubenfonseca !! Merging now, and will send one last PR to enable E2E at CI level and document it in the contributing guide.

**Pull request merged**